### PR TITLE
Feature/query operators

### PR DESCRIPTION
**Pull request by loudmouth**

### CODE REVIEW

File: Sources/Query.swift [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+            completeSelections.append("sys")
+        }
+        return completeSelections
+    }
+}
+
+extension String {
+
+    func isValidSelection() -> Bool {
+        if characters.split(separator: ".").count > 2 {
```
Comment (mariobodemann) [inline Sources/Query.swift]: `fields..foo` will still pass... What about `fields..foo".characters.split(separator:".", maxSplits:34, omittingEmptySubsequences:false)`

File: Sources/Sys.swift [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+    public init(map: Map) throws {
+        id              = try map.value("id")
+        type            = try map.value("type")
+
+        locale          = try? map.value("locale")
+        contentTypeId   = try? map.value("contentType.sys.id")
+        createdAt       = try? map.value("createdAt")
+        updatedAt       = try? map.value("updatedAt")
+    }
+}
```
Comment (mariobodemann) [inline Sources/Sys.swift]: Do you want to map `space` and `revision` too?

File: Tests/EntryTests.swift [hunk 3]
```diff
@@ -230,8 +228,14 @@ class EntryTests: XCTestCase {
...
                 expect(entry.fields["name"] as? String).to(equal("Nyan Cat"))
 
-                let image = entry.fields["image"] as? Asset
-                expect(url(image!).absoluteString).to(equal("https://images.contentful.com/cfexampleapi/4gp6taAwW4CmSgumq2ekUm/9da0cd1936871b8d72343e895a00d611/Nyan_cat_250px_frame.png"))
+                if let assetLink = entry.fields["image"] as? Link {
+                    switch assetLink {
+                    case .asset(let image):
+                        expect(url(image).absoluteString).to(equal("https://images.contentful.com/cfexampleapi/4gp6taAwW4CmSgumq2ekUm/9da0cd1936871b8d72343e895a00d611/Nyan_cat_250px_frame.png"))
+                    default:
+                        fail("Should not have a link of the wrong type her")
```
Comment (mariobodemann) [inline Tests/EntryTests.swift]: `here`?

File: Tests/LinkTests.swift [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+import Nimble
+import DVR
+
+class LinkTests: XCTestCase {
+
+    func testGetLinkFromFieldValueThatIsAlreadyALink() {
+        let entry = Entry(localizedField: ["":["":""]])
+        let link = Link.entry(entry)
+        let newLink = Link.link(from: link)
+        expect(newLink).toNot(beNil())
```
Comment (mariobodemann) [inline Tests/LinkTests.swift]: Can this test be more explicit? I.e. what about changing it from `.toNot(beNil())` to testing if the `newLink` sys.id equals the linked entry id, or so?

File: Tests/QueryTests.swift [hunk 5]
```diff
@@ -0,0 +1,343 @@
...
+@testable import Contentful
+import XCTest
+import Nimble
+import DVR
+import Interstellar
+import CoreData
+
+
+// PERSISTENCE
+//final class Cat: NSManagedObject, ContentModel {
```
Comment (mariobodemann) [inline Tests/QueryTests.swift]: Are those tests to be used once persistence is updated, or can we uncomment them, make them useful?

File: Tests/QueryTests.swift [hunk 6]
```diff
@@ -0,0 +1,343 @@
...
+
+        let expectation = self.expectation(description: "Select operator expectation")
+
+        let query = try! Query<Dog>.select(fieldNames: selections)
+
+        QueryTests.client.fetchEntries(with: query) { result in
+
+            switch result {
+            case .success(let dogs):
+                let doge = dogs.first!
```
Comment (mariobodemann) [inline Tests/QueryTests.swift]: Is it wise to depend on the undefined ordering here?

File: Tests/QueryTests.swift [hunk 7]
```diff
@@ -0,0 +1,343 @@
...
+        let fieldNames = ["sys.contentType.sys.id"]
+
+        do {
+            let _ = try Query<Dog>.select(fieldNames: fieldNames)
+
+            fail("Query selection with depth > 2 should throw an error and not reahc here")
+        } catch let error as QueryError {
+            expect(error.message).toNot(beNil())
+        } catch _ {
+            fail("Should throw a QueryError")
```
Comment (mariobodemann) [inline Tests/QueryTests.swift]: do you need to catch the exception here, or can you let it bubble through to the test?

File: Sources/Link.swift [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+//
+
+import Foundation
+import ObjectMapper
+
+public struct LinkSys {
+
+    public let id: String
+
+    public let linkType: String // "Entry" or "Asset" (-> Easier to resolve with this information.
```
Comment (loudmouth) [inline Sources/Link.swift]: I'll just rm this comment. It was just for me ;-)

File: Tests/LinkTests.swift (hunk 4 above)
Comment (loudmouth) [inline Tests/LinkTests.swift]: ooops, I thought I had deleted this file. Turns out i only removed it from the xcode project index. It's not a necessary test.

File: Tests/QueryTests.swift (hunk 6 above)
Comment (loudmouth) [inline Tests/QueryTests.swift]: i'm not sure...any better ideas?

File: Tests/QueryTests.swift (hunk 7 above)
Comment (loudmouth) [inline Tests/QueryTests.swift]: I'm catching here to make sure that the error thrown is the right type of `Error`. So yes, I do indeed need to catch it now.

File: Tests/QueryTests.swift (hunk 6 above)
Comment (mariobodemann) [inline Tests/QueryTests.swift]: Yes: Also request an ordering by any field ... ;)

File: Tests/QueryTests.swift (hunk 5 above)
Comment (loudmouth) [inline Tests/QueryTests.swift]: oops. i forgot to address those. I'll remove them

**Pull request merged**